src/hotspot/share/opto/node.cpp

```diff
@@ -894,33 +894,30 @@ int Node::replace_edges_in_range(Node* old, Node* neww, int start, int end) {
 
 //-------------------------disconnect_inputs-----------------------------------
 // NULL out all inputs to eliminate incoming Def-Use edges.
-// Return the number of edges between 'n' and 'this'
-int Node::disconnect_inputs(Node *n, Compile* C) {
-  int edges_to_n = 0;
-
-  uint cnt = req();
-  for( uint i = 0; i < cnt; ++i ) {
-    if( in(i) == 0 ) continue;
-    if( in(i) == n ) ++edges_to_n;
-    set_req(i, NULL);
+void Node::disconnect_inputs(Compile* C) {
+  // r: a required input, null is allowed
+  // p: a precedence, null values are all at the end
+  // |r|...|r|p|...|p|null|...|null|
+  //         |                     |
+  //         req()                 len()
+  // -----------------------------------
+  uint i = 0;
+
+  for (; i < req(); ++i) {
+    if(in(i) == nullptr) continue;
+    set_req(i, nullptr);
   }
+
   // Remove precedence edges if any exist
   // Note: Safepoints may have precedence edges, even during parsing
-  if( (req() != len()) && (in(req()) != NULL) ) {
-    uint max = len();
-    for( uint i = 0; i < max; ++i ) {
-      if( in(i) == 0 ) continue;
-      if( in(i) == n ) ++edges_to_n;
-      set_prec(i, NULL);
-    }
+  // Precedences have no embedded NULL
+  for (; i < len() && in(i) != nullptr; ++i) {
```

Comment:
I don't think it is correct.

Reply:
hi, @cl4es  @vnkozlov, 
Thank you to review this patch. 

The reason I changed to this because I read this comment yesterday. https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/node.hpp#L280

It suggests that `Node::_in` has an inherent order, which I depict here: 
https://github.com/navyxliu/jdk/commit/b8a72755a29d2fcbe36c9dcaf7f696634f813b4f#diff-e1c5a771a82d5fdb7e88c5b90b444736R898

If C2 sees a nullptr in precedence section, it can assume the rest of precedences are all nullptr. 
I have 2 evidence to prove it holds. 
1. rm_prec() does relocate null value to the end. https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/node.cpp#L1025
2. I added an assert after the precedence loop and verified it using hotspot-tier1.  
`for(; i<len(); ++i) assert(in(i) == nullptr, "non-null prec?");
`

With this assumption, c2 can leave loop a little bit early. 

How about this? I withdraw this change and just make a pure clean-up. 
I admit that the potential gain is very very small.  it's not worth it.

Reply:
You are right. There is already existing assert in find_prec_edge() called from set_prec():
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/node.hpp#L436

I still want to see explicit start from req():

```
for (uint i = req(); (i < len()) && (in(i) != nullptr); ++i) {
  set_prec(i, nullptr);
}
```